File: humchess/inference.py
"""
Inference module for move prediction.

Provides functions to predict moves given a position, Elo, and time.
"""

from typing import Optional

import chess
import torch
import torch.nn.functional as F

from .data.tokenization import (
    NUM_MOVE_CLASSES,
    fen_to_tokens, normalize_position, move_to_ids, ids_to_move,
    is_promotion_move, denormalize_move,
)
from .model.transformer import ChessTransformer


def get_legal_move_mask(board: chess.Board, was_black: bool) -> torch.Tensor:
    """
    Create legal move mask for a position.

    Args:
        board: Chess board (in original orientation)
        was_black: Whether original position was black to move

    Returns:
        Boolean tensor of shape (4096,) - True for legal moves
    """
    mask = torch.zeros(NUM_MOVE_CLASSES, dtype=torch.bool)

    for move in board.legal_moves:
        from_sq = move.from_square
        to_sq = move.to_square

        if was_black:
            from_sq = 63 - from_sq
            to_sq = 63 - to_sq

        move_id = from_sq * 64 + to_sq
        mask[move_id] = True

    return mask


@torch.no_grad()
def predict_move(
    model: ChessTransformer,
    fen: str,
    elo: int,
    time_left_seconds: Optional[float] = None,
    device: torch.device = None,
) -> str:
    """
    Predict a move for a given position.

    Args:
        model: Trained ChessTransformer model
        fen: FEN string of the position
        elo: Elo rating to emulate
        time_left_seconds: Time remaining on clock (optional)
        device: Device to run inference on

    Returns:
        UCI move string (e.g., 'e2e4', 'e7e8q')
    """
    if device is None:
        device = next(model.parameters()).device

    model.eval()

    # Parse position
    board = chess.Board(fen)

    # Convert to tokens (returns is_black flag)
    tokens, is_black = fen_to_tokens(fen, elo, time_left_seconds)

    # Create a dummy move for normalization (we'll use any legal move)
    dummy_move = list(board.legal_moves)[0].uci()
    tokens, _ = normalize_position(tokens, dummy_move, is_black)

    # Create input tensor
    tokens_tensor = torch.tensor(tokens, dtype=torch.long, device=device).unsqueeze(0)

    # Get legal move mask
    legal_mask = get_legal_move_mask(board, is_black).to(device).unsqueeze(0)

    # Forward pass
    outputs = model(tokens_tensor)
    move_logits = outputs['move_logits']  # (1, 4096)
    promo_logits = outputs['promo_logits']  # (1, 4)

    # Apply legality mask
    masked_logits = move_logits.clone()
    masked_logits[~legal_mask] = float('-inf')

    # Sample from distribution
    probs = F.softmax(masked_logits, dim=-1)
    move_id = torch.multinomial(probs, num_samples=1).item()

    # Check if promotion
    board_tokens = tokens[1:65]
    is_promo = is_promotion_move(move_id, board_tokens)

    promo_id = None
    if is_promo:
        promo_probs = F.softmax(promo_logits, dim=-1)
        promo_id = torch.multinomial(promo_probs, num_samples=1).item()

    # Convert to UCI
    move_uci = ids_to_move(move_id, promo_id)

    # Denormalize if needed
    move_uci = denormalize_move(move_uci, is_black)

    return move_uci


@torch.no_grad()
def predict_move_distribution(
    model: ChessTransformer,
    fen: str,
    elo: int,
    time_left_seconds: Optional[float] = None,
    device: torch.device = None,
    top_k: int = 5,
) -> list[tuple[str, float]]:
    """
    Get top-k moves with their probabilities.

    Args:
        model: Trained ChessTransformer model
        fen: FEN string of the position
        elo: Elo rating to emulate
        time_left_seconds: Time remaining on clock (optional)
        device: Device to run inference on
        top_k: Number of top moves to return

    Returns:
        List of (uci_move, probability) tuples, sorted by probability descending
    """
    if device is None:
        device = next(model.parameters()).device

    model.eval()

    # Parse position
    board = chess.Board(fen)

    # Convert to tokens (returns is_black flag)
    tokens, is_black = fen_to_tokens(fen, elo, time_left_seconds)

    # Dummy move for normalization
    dummy_move = list(board.legal_moves)[0].uci()
    tokens, _ = normalize_position(tokens, dummy_move, is_black)

    # Create input tensor
    tokens_tensor = torch.tensor(tokens, dtype=torch.long, device=device).unsqueeze(0)

    # Get legal move mask
    legal_mask = get_legal_move_mask(board, is_black).to(device).unsqueeze(0)

    # Forward pass
    outputs = model(tokens_tensor)
    move_logits = outputs['move_logits']  # (1, 4096)

    # Apply legality mask
    masked_logits = move_logits.clone()
    masked_logits[~legal_mask] = float('-inf')

    # Get probabilities
    probs = F.softmax(masked_logits, dim=-1).squeeze(0)  # (4096,)

    # Get top-k
    top_probs, top_ids = torch.topk(probs, k=min(top_k, legal_mask.sum().item()))

    # Convert to moves
    results = []
    for move_id, prob in zip(top_ids.tolist(), top_probs.tolist()):
        # For simplicity, assume queen promotion for display
        # (full version would show all promotion options for pawn moves)
        board_tokens = tokens[1:65]
        is_promo = is_promotion_move(move_id, board_tokens)
        promo_id = 0 if is_promo else None  # Queen promotion

        move_uci = ids_to_move(move_id, promo_id)
        move_uci = denormalize_move(move_uci, is_black)
        results.append((move_uci, prob))

    return results


def load_model(checkpoint_path: str, device: torch.device = None) -> ChessTransformer:
    """
    Load a trained model from checkpoint.

    Args:
        checkpoint_path: Path to checkpoint file
        device: Device to load model to

    Returns:
        Loaded ChessTransformer model
    """
    if device is None:
        device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')

    checkpoint = torch.load(checkpoint_path, map_location=device)

    # Infer model size from state dict
    state_dict = checkpoint['model_state_dict']
    d_model = state_dict['token_embedding.weight'].shape[1]
    n_layers = sum(1 for k in state_dict if 'self_attn.in_proj_weight' in k)

    # Create model with matching architecture
    # This is a heuristic - ideally save config in checkpoint
    if d_model == 128:
        model = ChessTransformer(d_model=128, n_heads=4, n_layers=n_layers, d_ff=512)
    elif d_model == 256:
        model = ChessTransformer(d_model=256, n_heads=8, n_layers=n_layers, d_ff=1024)
    elif d_model == 512:
        model = ChessTransformer(d_model=512, n_heads=8, n_layers=n_layers, d_ff=2048)
    else:
        model = ChessTransformer(d_model=d_model, n_layers=n_layers)

    model.load_state_dict(state_dict)
    model = model.to(device)
    model.eval()

    return model

File: web-inference/static/js/inference.js
/**
 * ONNX Runtime Web inference for HumChess.
 *
 * Handles model loading and move prediction.
 */

import {
    boardToTokens,
    normalizePosition,
    denormalizeMove,
    idToMove,
    isPromotionMove,
    NUM_MOVE_CLASSES,
    Piece,
} from './tokenization.js';

// Global inference session
let session = null;

/**
 * Load the ONNX model.
 * @param {string} modelPath - Path to ONNX model file
 * @param {function} onProgress - Progress callback (0-1)
 * @returns {Promise<void>}
 */
export async function loadModel(modelPath, onProgress = null) {
    if (onProgress) onProgress(0);

    // Configure ONNX Runtime
    // WASM backend supports int64 (required for embeddings)
    const options = {
        executionProviders: ['wasm'],
        graphOptimizationLevel: 'all',
    };

    try {
        session = await ort.InferenceSession.create(modelPath, options);
        console.log('Model loaded successfully');
        console.log('Input names:', session.inputNames);
        console.log('Output names:', session.outputNames);
        if (onProgress) onProgress(1);
    } catch (error) {
        console.error('Failed to load model:', error);
        throw error;
    }
}

/**
 * Check if model is loaded.
 * @returns {boolean}
 */
export function isModelLoaded() {
    return session !== null;
}

/**
 * Get legal move mask from chess.js instance.
 * @param {Chess} chess - chess.js instance
 * @param {boolean} isBlackToMove - Whether original position was black to move
 * @returns {Float32Array} - Mask of shape (4096,), 0 for legal, -Infinity for illegal
 */
function getLegalMoveMask(chess, isBlackToMove) {
    const mask = new Float32Array(NUM_MOVE_CLASSES).fill(-Infinity);

    const moves = chess.moves({ verbose: true });

    for (const move of moves) {
        let fromSq = squareToIdx(move.from);
        let toSq = squareToIdx(move.to);

        // Apply normalization if black to move
        if (isBlackToMove) {
            fromSq = 63 - fromSq;
            toSq = 63 - toSq;
        }

        const moveId = fromSq * 64 + toSq;
        mask[moveId] = 0;
    }

    return mask;
}

/**
 * Convert chess.js square notation to index.
 * @param {string} sq - Square like "e4"
 * @returns {number}
 */
function squareToIdx(sq) {
    const file = sq.charCodeAt(0) - 'a'.charCodeAt(0);
    const rank = parseInt(sq[1]) - 1;
    return rank * 8 + file;
}

/**
 * Apply softmax to logits.
 * @param {Float32Array} logits
 * @returns {Float32Array}
 */
function softmax(logits) {
    const maxLogit = Math.max(...logits.filter(x => isFinite(x)));
    const exps = logits.map(x => isFinite(x) ? Math.exp(x - maxLogit) : 0);
    const sum = exps.reduce((a, b) => a + b, 0);
    return exps.map(x => x / sum);
}

/**
 * Sample from a probability distribution.
 * @param {Float32Array} probs - Probability distribution
 * @returns {number} - Sampled index
 */
function sampleFromDistribution(probs) {
    const r = Math.random();
    let cumsum = 0;
    for (let i = 0; i < probs.length; i++) {
        cumsum += probs[i];
        if (r < cumsum) {
            return i;
        }
    }
    return probs.length - 1;
}

/**
 * Get the best move (argmax) from logits.
 * @param {Float32Array} logits - Masked logits
 * @returns {number} - Best move index
 */
function argmax(logits) {
    let bestIdx = 0;
    let bestVal = logits[0];
    for (let i = 1; i < logits.length; i++) {
        if (logits[i] > bestVal) {
            bestVal = logits[i];
            bestIdx = i;
        }
    }
    return bestIdx;
}

/**
 * Predict the next move.
 *
 * @param {Chess} chess - chess.js instance with current position
 * @param {number} elo - Target Elo rating for the AI
 * @param {number|null} timeLeftSeconds - Time remaining (optional)
 * @param {Object} options - Options
 * @param {boolean} options.sample - Whether to sample (true) or take argmax (false)
 * @param {number} options.temperature - Sampling temperature (only if sample=true)
 * @returns {Promise<{move: string, moveId: number, promoId: number|null, probs: Float32Array}>}
 */
export async function predictMove(chess, elo, timeLeftSeconds = null, options = {}) {
    if (!session) {
        throw new Error('Model not loaded. Call loadModel() first.');
    }

    const { sample = true, temperature = 1.0 } = options;

    // Convert board to tokens
    const { tokens, isBlackToMove } = boardToTokens(chess, elo, timeLeftSeconds);

    // Apply white normalization
    const normalizedTokens = normalizePosition(tokens, isBlackToMove);

    // Get legal move mask
    const legalMask = getLegalMoveMask(chess, isBlackToMove);

    // Create input tensor
    const inputTensor = new ort.Tensor(
        'int64',
        BigInt64Array.from(normalizedTokens.map(BigInt)),
        [1, 68]
    );

    // Run inference
    const startTime = performance.now();
    const outputs = await session.run({ tokens: inputTensor });
    const inferenceTime = performance.now() - startTime;

    // Get logits
    const moveLogits = outputs.move_logits.data;
    const promoLogits = outputs.promo_logits.data;

    // Apply legal mask
    const maskedLogits = new Float32Array(NUM_MOVE_CLASSES);
    for (let i = 0; i < NUM_MOVE_CLASSES; i++) {
        maskedLogits[i] = moveLogits[i] + legalMask[i];
    }

    // Apply temperature and get probabilities
    let scaledLogits = maskedLogits;
    if (temperature !== 1.0) {
        scaledLogits = maskedLogits.map(x => x / temperature);
    }
    const probs = softmax(scaledLogits);

    // Select move
    let moveId;
    if (sample) {
        moveId = sampleFromDistribution(probs);
    } else {
        moveId = argmax(maskedLogits);
    }

    // Check for promotion
    const boardTokens = normalizedTokens.slice(1, 65);
    const isPromotion = isPromotionMove(moveId, boardTokens);

    let promoId = null;
    if (isPromotion) {
        const promoProbs = softmax(Array.from(promoLogits));
        if (sample) {
            promoId = sampleFromDistribution(promoProbs);
        } else {
            promoId = argmax(Array.from(promoLogits));
        }
    }

    // Convert to UCI move
    let move = idToMove(moveId, promoId);

    // Denormalize if needed
    move = denormalizeMove(move, isBlackToMove);

    return {
        move,
        moveId,
        promoId,
        probs,
        inferenceTime,
        isBlackToMove,
    };
}

/**
 * Get move probabilities without sampling.
 *
 * @param {Chess} chess - chess.js instance
 * @param {number} elo - Target Elo
 * @param {number|null} timeLeftSeconds - Time remaining
 * @returns {Promise<{moves: Array<{move: string, prob: number}>, inferenceTime: number}>}
 */
export async function getMoveDistribution(chess, elo, timeLeftSeconds = null) {
    if (!session) {
        throw new Error('Model not loaded. Call loadModel() first.');
    }

    const { tokens, isBlackToMove } = boardToTokens(chess, elo, timeLeftSeconds);
    const normalizedTokens = normalizePosition(tokens, isBlackToMove);
    const legalMask = getLegalMoveMask(chess, isBlackToMove);

    const inputTensor = new ort.Tensor(
        'int64',
        BigInt64Array.from(normalizedTokens.map(BigInt)),
        [1, 68]
    );

    const startTime = performance.now();
    const outputs = await session.run({ tokens: inputTensor });
    const inferenceTime = performance.now() - startTime;

    const moveLogits = outputs.move_logits.data;

    // Apply mask and softmax
    const maskedLogits = new Float32Array(NUM_MOVE_CLASSES);
    for (let i = 0; i < NUM_MOVE_CLASSES; i++) {
        maskedLogits[i] = moveLogits[i] + legalMask[i];
    }
    const probs = softmax(maskedLogits);

    // Get all legal moves with their probabilities
    const legalMoves = chess.moves({ verbose: true });
    const moveProbs = [];

    for (const legalMove of legalMoves) {
        let fromSq = squareToIdx(legalMove.from);
        let toSq = squareToIdx(legalMove.to);

        if (isBlackToMove) {
            fromSq = 63 - fromSq;
            toSq = 63 - toSq;
        }

        const moveId = fromSq * 64 + toSq;
        const prob = probs[moveId];

        // Get the actual UCI move (with promotion if applicable)
        let uci = legalMove.from + legalMove.to;
        if (legalMove.promotion) {
            uci += legalMove.promotion;
        }

        moveProbs.push({
            move: uci,
            san: legalMove.san,
            prob: prob,
        });
    }

    // Sort by probability descending
    moveProbs.sort((a, b) => b.prob - a.prob);

    return {
        moves: moveProbs,
        inferenceTime,
    };
}
